ls1intum/Artemis
# Add caching service worker to become an installable Progressive Web App

### Checklist

### Motivation and Context
The so called PWAs (Progressive Web Apps) nowadays allows users to install an website on their operating system as an native-ish app. By implementing a service-worker with a fitting manifest, the browser is able to install the application and even to gain access to operating system specific features like the notification system of e.g. Android, IOS or Windows. 
Service workers come with the potential to use several caching methodologies to design an offline experience for the users. Solving text or quiz exercises and even offline exams could become possible (like proposed here  #1572 ).

### Description
In the webpack.commons file I used the workbox plugin of jhipsters webpack and extended the existing GenerateSW class with the new required properties to enable runtime caching. 
A cache called "artemisCache" now caches all css, scss, html and js files for 24h using the [StaleWhileRevalidate](https://developers.google.com/web/tools/workbox/reference-docs/latest/module-workbox-strategies#stalewhilerevalidate) method. Alternately NetworkFirst could be used instead. This has the be tested on a test server.
An other cache called "assetCache" caches all static files like images or fonts waiting for an update of a newer version also using the StaleWhileRevalidate method.
For the installation on different devices its required to serve different resolutions of the logo. Therefore I generated four different resolutions based on the existing one, added them to the client images folder and linked them in the manifest file.
The registration of the service-worker is added inside the script tag of the index.html file.
While local testing I ran into memory shortage, which lead me to increase the client server to 4GB.
Offline or notification support isnt implemented yet.

### Steps for Testing

1. Open the website using a pwa compatible browser like Chrome, Edge or (maybe) Safari (Firefox can not install the app!)
2. After the site loaded completely a notice from the browser will occur that the website could be installed. Some browsers place that on the right in the url bar. 
3. Install the app
4. It should now be installed as a normal application on the operating system. Try search and open it
5. Artemis should now be usable as usual

### What still needs to be tested
- [x] Chrome support on windows
- [ ] Safari support
- [ ] Edge support
- [ ] Chrome on mobile support
- [ ] webpack.common doesnt collide with the webpack.prod GenerateSW implementation
- [ ] StaleWhileRevalidate works without problems on artemisCache

### Screenshots

#### Installation notice
![image](https://user-images.githubusercontent.com/33132009/115160181-3256fe80-a097-11eb-9fbc-e3533dfabc52.png)

#### Installed App on Windows
![image](https://user-images.githubusercontent.com/33132009/115160191-3f73ed80-a097-11eb-954a-5ce81d3f9902.png)
